File: src/main/java/yuma140902/uptodatemod/blocks/BlockCoarseDirt.java
package yuma140902.uptodatemod.blocks;

import cpw.mods.fml.common.eventhandler.Event.Result;
import cpw.mods.fml.common.registry.GameRegistry;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.util.IIcon;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.common.IPlantable;
import net.minecraftforge.common.util.ForgeDirection;
import net.minecraftforge.event.entity.player.UseHoeEvent;
import yuma140902.uptodatemod.IHasRecipes;
import yuma140902.uptodatemod.IRegisterable;
import yuma140902.uptodatemod.MyBlocks;
import yuma140902.uptodatemod.util.StringUtil;

public class BlockCoarseDirt extends Block implements IRegisterable, IHasRecipes {

	public static void onUseHoeEvent(UseHoeEvent event) {
		if(event.world.getBlock(event.x, event.y, event.z) == MyBlocks.coarseDirt) {
			event.world.setBlock(event.x, event.y, event.z, Blocks.dirt);
			event.world.playSoundEffect(event.x + 0.5F, event.y + 0.5F, event.z + 0.5F, Block.soundTypeGravel.getStepResourcePath(), 1.0F, 0.8F);
			event.setResult(Result.ALLOW);
		}
	}
	
	
	public BlockCoarseDirt() {
		super(Material.ground);
		setHardness(0.5F);
		setStepSound(soundTypeGravel);
		setHarvestLevel("shovel", 0);
		setCreativeTab(CreativeTabs.tabBlock);
	}
	
	@Override
	public void register() {
		this.setBlockName(StringUtil.getDomainedUnlocalizedName("coarse_dirt"));
		this.setBlockTextureName(StringUtil.getDomainedTextureName("coarse_dirt"));
		GameRegistry.registerBlock(this, "coarse_dirt");
	}
	
	@Override
	public boolean
			canSustainPlant(IBlockAccess world, int x, int y, int z, ForgeDirection direction, IPlantable plantable) {
		return Blocks.dirt.canSustainPlant(world, x, y, z, direction, plantable);
	}
	
	@Override
	@SideOnly(Side.CLIENT)
	public IIcon getIcon(IBlockAccess world, int x, int y, int z, int p_149673_5_) {
		Material material = world.getBlock(x, y + 1, z).getMaterial();

    if (material == Material.snow || material == Material.craftedSnow)
    {
        return Blocks.grass.getIcon(world, x, y, z, p_149673_5_);
    }
		
		return super.getIcon(world, x, y, z, p_149673_5_);
	}
	
	@Override
	public void registerRecipes() {
		GameRegistry.addRecipe(
				new ItemStack(MyBlocks.coarseDirt, 4, 0),
				"GD",
				"DG",
				'D', Blocks.dirt,
				'G', Blocks.gravel
				);
		
		GameRegistry.addRecipe(
				new ItemStack(MyBlocks.coarseDirt, 4, 0),
				"DG",
				"GD",
				'D', Blocks.dirt,
				'G', Blocks.gravel
				);
	}
	
}

File: src/main/scala/yuma140902/uptodatemod/blocks/BlockCoarseDirt.scala
package yuma140902.uptodatemod.blocks

import yuma140902.uptodatemod.registry.RecipeRegister
import yuma140902.uptodatemod.{ModUpToDateMod, MyBlocks}
import yuma140902.yumalib.api.util.Name
import yuma140902.yumalib.api.util.NameExtensions._
import yuma140902.yumalib.api.{IHasRecipes, RegisterableBlock}

import cpw.mods.fml.common.eventhandler.Event.Result
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.init.Blocks
import net.minecraft.item.{ItemBlock, ItemStack}
import net.minecraft.util.IIcon
import net.minecraft.world.IBlockAccess
import net.minecraftforge.common.IPlantable
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.event.entity.player.UseHoeEvent

// TODO: object
class BlockCoarseDirt extends Block(Material.ground) with RegisterableBlock[ItemBlock] with IHasRecipes {
  override val name: Name = Name("coarse_dirt")

  this.setHardness(0.5F)
  this.setStepSound(Block.soundTypeGravel)
  this.setHarvestLevel("shovel", 0)
  this.setCreativeTab(CreativeTabs.tabBlock)
  this.setBlockTextureName(ModUpToDateMod.name.domainedTexture(name))

  override def canSustainPlant(world: IBlockAccess, x: Int, y: Int, z: Int, direction: ForgeDirection, plantable: IPlantable): Boolean =
    Blocks.dirt.canSustainPlant(world, x, y, z, direction, plantable)

  @SideOnly(Side.CLIENT)
  override def getIcon(world: IBlockAccess, x: Int, y: Int, z: Int, side: Int): IIcon = {
    val material = world.getBlock(x, y + 1, z).getMaterial
    if ((material eq Material.snow) || (material eq Material.craftedSnow)) return Blocks.grass.getIcon(world, x, y, z, side)
    super.getIcon(world, x, y, z, side)
  }


  override def registerRecipes(): Unit = {

    RecipeRegister.addShaped(
      new ItemStack(MyBlocks.coarseDirt, 4, 0),
      "GD",
      "DG",
      'D'.asInstanceOf[Object], Blocks.dirt,
      'G'.asInstanceOf[Object], Blocks.gravel
    )

    RecipeRegister.addShaped(
      new ItemStack(MyBlocks.coarseDirt, 4, 0),
      "DG",
      "GD",
      'D'.asInstanceOf[Object], Blocks.dirt,
      'G'.asInstanceOf[Object], Blocks.gravel
    )
  }
}

object BlockCoarseDirt {
  def onUseHoeEvent(event: UseHoeEvent): Unit = {
    if (event.world.getBlock(event.x, event.y, event.z) eq MyBlocks.coarseDirt) {
      event.world.setBlock(event.x, event.y, event.z, Blocks.dirt)
      event.world.playSoundEffect(event.x + 0.5F, event.y + 0.5F, event.z + 0.5F, Block.soundTypeGravel.getStepResourcePath, 1.0F, 0.8F)
      event.setResult(Result.ALLOW)
    }
  }
}